spec/features/preview_article_edit_spec.rb

```diff
@@ -0,0 +1,21 @@
+require 'rails_helper'
+
+RSpec.feature "Preview Article Changes", type: :feature, js: true do
+  let(:admin){ FactoryGirl.create(:admin_user) }
+  let(:article) { FactoryGirl.create(:article) }
+  before { login(admin) }
+
+  scenario "admin can preview article changes" do
+    new_name = "Edited name"
+    visit edit_admin_article_path(article)
+    fill_in "Name", with: new_name
+    click_on "Preview" 
+
+    sleep(5)
```

Comment:
I tried implementing `wait_until`, but was unable to get it to work -- kept getting timeout errors. This was how it was implemented (slight update to timeout syntax based on docs):
```ruby
def wait_until                                                                                                                                                                                                                             
  require "timeout"                                                              
  Timeout::timeout(5) do                                                         
    sleep(0.1) until value = yield                                               
    value                                                                        
  end                                                                            
end  
```
and this was how I used it in the tests instead of `sleep`
```ruby
wait_until { page.driver.browser.window_handles == 2 }
```
...but it seemed to never exit `wait_until`. I did find that it only needs to sleep for 1 second for the window to load, so I've changed it to that for now. Do you know what could be going on with `wait_until`? It seems like a better solution

Reply:
I looked at this a bit on Friday. It looks seems like it's successfully waiting until the new window appears and switching to it, but eventually throwing a read timeout error when it tries to actually read from the new window (ie the code fails on line 18). In other words, it seems like it needs to wait until a little *after* the new window appears in the `window_handles` array before attempting to read.

I found a wrapper in Capybara for this and it still seems to have the problem of attempting to switch too early. So, this works:
```
switch_to_window window_opened_by { click_on 'Preview' }
sleep 1
# Proceed with actions
```
but removing the sleep causes the read timeout error.

I'm not sure why this happens! I think it's time to throw a `wait` in there.